packages/manganis-core/src/images.rs

```diff
@@ -0,0 +1,177 @@
+use const_serialize::SerializeConst;
+
+use crate::AssetOptions;
+
+/// The type of an image. You can read more about the tradeoffs between image formats [here](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types)
+#[derive(
+    Debug,
+    PartialEq,
+    PartialOrd,
+    Clone,
+    Copy,
+    Hash,
+    SerializeConst,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+#[repr(u8)]
+pub enum ImageFormat {
+    /// A png image. Png images cannot contain transparency and tend to compress worse than other formats
+    Png,
+    /// A jpg image. Jpg images can contain transparency and tend to compress better than png images
+    Jpg,
+    /// A webp image. Webp images can contain transparency and tend to compress better than jpg images
+    Webp,
+    /// An avif image. Avif images can compress slightly better than webp images but are not supported by all browsers
+    Avif,
+    /// An unknown image type
+    Unknown,
+}
+
+/// The size of an image asset
+#[derive(
+    Debug,
+    PartialEq,
+    PartialOrd,
+    Clone,
+    Copy,
+    Hash,
+    SerializeConst,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+#[repr(C, u8)]
+pub enum ImageSize {
+    /// A manual size in pixels
+    Manual {
+        /// The width of the image in pixels
+        width: u32,
+        /// The height of the image in pixels
+        height: u32,
+    },
+    /// The size will be automatically determined from the image source
+    Automatic,
+}
+
+/// A builder for an image asset. This must be used in the [`mg!`] macro.
+#[derive(
+    Debug,
+    PartialEq,
+    PartialOrd,
+    Clone,
+    Copy,
+    Hash,
+    SerializeConst,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+pub struct ImageAssetOptions {
+    ty: ImageFormat,
+    low_quality_preview: bool,
+    size: ImageSize,
+    preload: bool,
+}
+
+impl Default for ImageAssetOptions {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ImageAssetOptions {
+    /// Create a new image asset options
+    pub const fn new() -> Self {
+        Self {
+            ty: ImageFormat::Unknown,
+            low_quality_preview: false,
+            size: ImageSize::Automatic,
+            preload: false,
+        }
+    }
+
+    /// Make the asset preloaded
+    ///
+    /// Preloading an image will make the image start to load as soon as possible. This is useful for images that will be displayed soon after the page loads or images that may not be visible immediately, but should start loading sooner
+    ///
+    /// ```rust
+    /// # use manganis::{asset, Asset, ImageAssetOptions};
+    /// const _: Asset = asset!("/assets/image.png", ImageAssetOptions::new().with_preload(true));
```

Comment:
ImageAsset doesn't exist. We can't support metadata about assets like lqip without injecting it in the linker, so I didn't add it in this PR because of the time constraint